docs/debugging.rst

```diff
@@ -0,0 +1,230 @@
+    use ``print`` in IDLE / Jupyter.
+
+.. note::
+
+    For the **OpenGL and CUDA backend**, the printed result won't shows up until ``ti.sync()``:
+
+    .. code-block:: python
+
+        import taichi as ti
+        ti.init(arch=ti.cuda)
+
+        @ti.kernel
+        def kern():
+            print('inside kernel')
+
+        print('before kernel')
+        kern()
+        print('after kernel')
+        ti.sync()
+        print('after sync')
+
+    obtains:
+
+    .. code-block:: none
+
+        before kernel
+        after kernel
+        inside kernel
+        after
+
+    Also note that host access or program end will also implicitly invoke for ``ti.sync()``.
+
+
+Compile-time ``ti.static_print``
+--------------------------------
+
+Sometimes it's also useful to print Python-scope objects / constants like data type, SNode.
+So, similar to ``ti.static`` we provide ``ti.static_print`` to print compile-time constants.
+It behave as same as Python-scope ``print`` does, just being embbed into Taichi kernel.
+
+.. code-block:: python
+
+    x = ti.var(ti.f32, (2, 3))
+
+    @ti.kernel
+    def inside_taichi_scope():
+        ti.static_print(x.shape)
+        # => (2, 3)
+        ti.static_print(x.data_type())
+        # => DataType.float32
+        for i in range(4):
+                ti.static_print(i.data_type())
+                # => DataType.int32
+                # will only print once
+
+Unlike ``print``, ``ti.static_print`` will only print the expression once at compile-time. And
+therefore has no runtime cost.
+
+
+Runtime ``assert`` in kernel
+----------------------------
+
+We may use ``assert`` statement in Taichi-scope. When assertion condition failed, a
+``RuntimeError`` will be raised to indicate error.
+
+To make ``assert`` work, first make sure you are using the **CPU backend**.
+For performance reason, ``assert`` is only work when ``debug`` mode is on, For example:
+
+.. code-block:: python
+
+    ti.init(arch=ti.cpu, debug=True)
+
+    x = ti.var(ti.f32, 128)
+
+    @ti.kernel
+    def do_sqrt_all():
+        for i in x:
+            assert x[i] >= 0
+            x[i] = ti.sqrt(x)
+
+
+When your debugging work is done, set ``debug=False``. Now ``assert`` will be simply ignored
+therefore no runtime overhead, making your program easy to debug without losing performance.
+
+
+Compile-time ``ti.static_assert``
+---------------------------------
+
+Like ``ti.static_print``, we also provide a static version for ``assert``:
+``ti.static_assert``, it can be useful to make assertion on data type / dimention / shape.
+It works no matter wheater ``debug=True`` is specified. When assertion failure, it will
+raise ``AssertionError`` as a Python-scope ``assert`` does.
+
+For example:
+
+.. code-block:: python
+
+    @ti.func
+    def is_odd(x: ti.template()):
+        ti.static_assert(x.data_type() == ti.i32, "is_odd() is only supported for i32")
+        return x % 2 == 1
+
```

Comment:
We may give ``ti.static_assert`` a function demonstration here? so that people can realize it receive 2 parameters, the first one is assert expresstion and the second is a statement.

Reply:
I thought all Python users who know `assert` knows this failure message feature?